File: Binary-Tree-BFS/binary-tree-right-side-view.py
from collections import deque


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        # Empty -> return early
        if not root:
            return []

        # Use deque for BFS
        q = deque()
        q.append(root)

        res = []  # res arr to store final results

        # level order traversal - we go layer by layer
        while q:
            res.append(q[-1].val)  # Add rightmost element form current level

            # Go through the current's level nodes and add its children from left to right. (It will be done on a "snapshot" of q length, meaning if we add elements to q inside loop it wont affect how many times the inner loop will execute)
            for _ in range(len(q)):
                popped = q.popleft()

                if popped.left:
                    q.append(popped.left)
                if popped.right:
                    q.append(popped.right)

        return res

File: Binary-Tree-BFS/binary-tree-right-side-view.cpp
#include <deque>
#include <vector>

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if(!root)
            return {};

        std::deque<TreeNode*> q;
        q.push_back(root);

        std::vector<int> res;

        while(!q.empty())
        {
            int level_size = q.size(); // C++: size()

            for(int i = 0 ; i < level_size ; ++i)
            {
                TreeNode* node = q.front(); // must read first
                q.pop_front();              // then remove

                if(i == level_size - 1)
                    res.push_back(node->val);

                if(node->left)
                    q.push_back(node->left);

                if(node->right)
                    q.push_back(node->right);
            }
        }

        return res;
    }
};
